src/api/r.c

```diff
@@ -0,0 +1,331 @@
+#endif
+
+#include <Rembedded.h>
+#include <Rinterface.h>
+#include <Rconfig.h>
+#include <Rdefines.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "api/renv.h"
+
+bool R_initialized_once = false;
+static bool initR(tic_mem *tic, const char *code);
+
+void evalR(tic_mem *memory, const char *code) {
+	setEnvironmentVariablesIfUnset();
+	if (!R_initialized_once) {
+		initR(memory, code);
+		R_initialized_once = true;
+	}
+	SEXP RESULT;
+	if (R_initialized_once) RESULT = Rf_eval(Rf_mkString(code), R_GlobalEnv);
+	#if defined ebug /* -DEBUG=ON */
+	Rprintf("%s", RESULT);
+	#endif
+}
+SEXP r_trace(tic_mem *memory, SEXP *args) {
+  const char *text = "Hello, woRld!";
+	u8 color;
+	((tic_core *) memory)->api.trace(memory, text, color ? color : 15);
+	return R_NilValue;
+}
+
+extern int R_running_as_main_program;   /* location within The R sources: ../unix/system.c */
+
+void R_CleanUp(Rboolean saveact, int status, int RunLast) { ; }
+
+void R_Suicide(const char *message)
+{
+	char  pp[1024];
+	snprintf(pp, 1024, "Fatal error, but can't kermit: %s\nExecute me, please.\n", message);
+	R_ShowMessage(pp);
+	exit(1);
+
+	while(1);
+}
+
+static tic_mem *tic_memory_static;
+
+void R_ShowMessage(const char *s) {
+/* Always use the sixteenth color. */
+	((tic_core *) tic_memory_static)->api.trace(tic_memory_static, s, 15);
+}
+
+/* This function is called with code, which is the entirety of the studio
+ * editor's code buffer (i.e. the entire game code as one string). */
+static bool initR(tic_mem *tic, const char *code) {
+	tic_memory_static = tic;
+
+  tic_core *core;
+  if ((core = (((tic_core *) tic))->currentVM) != NULL) {
+  	Rf_endEmbeddedR(0);
+  	core->currentVM = NULL;
+  }
+
+	/* TODO: it makes sense to write the code to a file and initialize R with its
+	 * input taken from that file, I think. */
+
+	/* embdRAV: embedded R argument vector. */
+	static char *embdRAV[] = { "TIC-80", "--quiet", "--vanilla" };
+
+  /* Without this nothing in R will work. */
+	setEnvironmentVariablesIfUnset();
+  static bool R_Initialized = false;
+
+	R_running_as_main_program = 0;
+
+	if (!R_Initialized) {
+		R_Initialized = (bool) Rf_initEmbeddedR(sizeof(embdRAV)/sizeof(embdRAV[0]), embdRAV);
+		R_running_as_main_program = 0;
+		/* Declared in Rinterface.h, defined in Rf_initEmbeddedR. */
+		R_Interactive = false;
+	}
+
+	/* TODO: I think it makes sense to evaluate the code, then check if this
+	 * exists, then end the init function with evaluating the TIC-80 function. */
+	Rf_eval(Rf_mkString("if (exists(\"BOOT\") && is.function(BOOT) BOOT() else BOOT <- function() NULL;;"),
+					R_GlobalEnv);
+
+	tic_script *config = (tic_get_script(tic));
+	for (int kiwi = 0; kiwi <= config->api_keywordsCount; kiwi++) {
+#define defineRFn_(keyword) Rf_eval(Rf_mkString("t80."#keyword" <- function(...) .C(r_"#keyword", ...);"), R_GlobalEnv);
+#define doDefineRFunction(keyword) defineRFn_(keyword)
+		doDefineRFunction(config->api_keywords[kiwi]);
```

Comment:
Unfortunately, `#` is a preprocessor operation, not a runtime operation. This expands at compile time into

```c
Rf_eval(Rf_mkString("t80.""config->api_keywords[kiwi]"" <- function(...) .C(r_""config->api_keywords[kiwi]"", ...);"), R_GlobalEnv);;
```

I'm afraid you might have to use a temporary buffer with C string functions in order to achieve proper string concatenation.

Reply:
> I'm afraid you might have to use a temporary buffer with C string functions in order to achieve proper string concatenation.

Indeed. I realized that the last time I was working on this and I am nearly finished working on that. Thank you for pointing this out.